contrib/epee/src/wipeable_string.cpp

```diff
@@ -109,9 +115,112 @@ void wipeable_string::push_back(char c)
   buffer.back() = c;
 }
 
-void wipeable_string::pop_back()
+void wipeable_string::operator+=(char c)
+{
+  push_back(c);
+}
+
+void wipeable_string::operator+=(const std::string &s)
+{
+  CHECK_AND_ASSERT_THROW_MES(size() < std::numeric_limits<size_t>::max() - s.size(), "Appended data too large");
+  grow(size(), size() + s.size());
+  const char *ptr = s.c_str();
+  while (*ptr)
+    push_back(*ptr++);
+}
+
+void wipeable_string::operator+=(const epee::wipeable_string &s)
+{
+  CHECK_AND_ASSERT_THROW_MES(size() < std::numeric_limits<size_t>::max() - s.size(), "Appended data too large");
+  grow(size(), size() + s.size());
+  const char *ptr = s.data();
+  size_t sz = s.size();
+  while (sz--)
+    push_back(*ptr++);
+}
+
+void wipeable_string::operator+=(const char *s)
+{
+  size_t slen = strlen(s);
+  CHECK_AND_ASSERT_THROW_MES(size() < std::numeric_limits<size_t>::max() - slen, "Appended data too large");
+  grow(size(), size() + slen);
+  while (*s)
+    push_back(*s++);
+}
+
+void wipeable_string::append(const char *data, size_t len)
+{
+  CHECK_AND_ASSERT_THROW_MES(size() < std::numeric_limits<size_t>::max() - len, "Appended data too large");
+  grow(size(), size() + len);
+  while (len--)
+    push_back(*data++);
+}
+
+void wipeable_string::trim()
+{
+  size_t prefix = 0;
+  while (prefix < size() && data()[prefix] == ' ')
+    ++prefix;
+  if (prefix > 0)
+    memmove(buffer.data(), buffer.data() + prefix, size() - prefix);
+
+  size_t suffix = 0;
+  while (suffix < size()-prefix && data()[size() - 1 - prefix - suffix] == ' ')
+    ++suffix;
+
+  resize(size() - prefix - suffix);
+}
+
+void wipeable_string::split(std::vector<wipeable_string> &fields)
+{
+  fields.clear();
+  size_t len = size();
+  const char *ptr = data();
+  bool space = true;
+  while (len--)
+  {
+    const char c = *ptr++;
+    if (c != ' ')
+    {
+      if (space)
+        fields.push_back({});
+      fields.back().push_back(c);
+    }
+    space = c == ' ';
+  }
+}
+
+bool wipeable_string::parse_hexstr(wipeable_string &res)
+{
+  res.clear();
+  if (size() & 1)
+    return false;
+  const size_t len = size();
+  const char *d = data();
+  res.grow(0, len / 2);
+  static constexpr const char hex[] = "0123456789abcdef";
+  for (size_t i = 0; i < len; i += 2)
+  {
+    char c = tolower(d[i]);
+    const char *ptr0 = strchr(hex, c);
```

Comment:
You're asking to replace one line by 6, where the line is pretty obvious ?
I obviously don't care about EBCDIC. AFAIK only very old IBM iron is using this. Do more recent ones do ?

Reply:
This would replace more than one line - looks like a total of 4. And its definitely going to be faster, `strchr` is quite complex. The implementation you have could be the basis for a locale/encoding independent scheme, but is not safe for that either currently.

Reply:
Did you mean this ? Looks like 6 to 1. You could play tricks to make it less, yes. It's more complicated.

```
#if 1
    char c0 = d[i];
    if ( 'A' <= c0 && c0 <= 'F')
         c0 = c0 - 'A' + 10;
    else if ('a' <= c0 && c0 <= 'a')
        c0 = c0 - 'a' + 10;
    else if ('0' <= c0 && c0 <= '9')
        c0 = c0 - '0';
    else
        return false;
    char c1 = d[i+1];
    if ( 'A' <= c1 && c1 <= 'F')
         c1 = c1 - 'A' + 10;
    else if ('a' <= c1 && c1 <= 'a')
        c1 = c1 - 'a' + 10;
    else if ('0' <= c1 && c1 <= '9')
        c1 = c1 - '0';
    else
        return false;
    res.push_back((c0<<4) | c1);
#else
    char c = atolower(d[i]);
    const char *ptr0 = strchr(hex, c);
    if (!ptr0)
      return false;
    c = atolower(d[i+1]);
    const char *ptr1 = strchr(hex, c);
    if (!ptr1)
      return false;
    res.push_back(((ptr0-hex)<<4) | (ptr1-hex));
#endif
```

Reply:
This omits the `atolower` code that must be added for that block. You could switch it back to `::tolower` and still make the claim of less lines of code (but again the asymmetry between locale based `tolower` and the `strchr` method that is a defacto utf8 only).

Either way, a nibble function could be written instead of the copy and paste. It will be inlined in most situations. And the `strchr` is hiding some of the complexity - that function requires a generic processing loop.  I'm not particularly concerned either way.